2-planning/2-mocking.adoc

```diff
@@ -0,0 +1,137 @@
+fake API before building the real one. The basic concepts were mentioned in the
+introduction, but primarily this is about getting feedback early, allowing you
+to tweak and change your way to a solution which is hopefully well suited for
+your clients. Folks scoff at this and say "You can never get things perfect",
+but you can absolutely cut out a lot back and forth and dodge a few daft
+mistakes by putting in some time first.
+
+The term "mock" for a lot of developers will have unit-testing connotations. In
+unit-testing, a mock is a fake implementation of a class or function, which
+accepts the same arguments as the real thing. It might return something pretty
+similar to expected output, and different test cases might even modify those
+returns to see how the code under test works.
+
+This is almost exactly the concept here, just at a HTTP level instead. This is
+done using a "mock server", which will respond to the expected endpoints, error
+for non-existent endpoints, often even provide realistic validation errors if a
+client sends it a nonsense request.
+
+== The Most Basic Mock Server EVER
+
+There are a lot of ways to create a mock server, but the most simple mock server
+around is https://github.com/typicode/json-server[JSON Server]. This thing is
+very handy for quickly getting a HTTP server spitting out some JSON that you
+control:
+
+.Woah what a surprise, another NPM tool!
+[source]
+----
+$ npm install -g json-server
+----
+
+Then you create a `db.json` file with some data in there:
+
+[source,javascript]
+----
+{
+  "posts": [
+    { "id": 1, "title": "json-server", "author": "typicode" }
+  ],
+  "comments": [
+    { "id": 1, "body": "some comment", "postId": 1 }
+  ],
+  "profile": { "name": "typicode" }
+}
+----
+
+Now, get that server started via the CLI.
+
+[source]
+----
+$ json-server --watch db.json
+----
+
+When browsing to `http://localhost:3000/posts/1` you will see the following JSON:
+
+[source]
+----
+{ "id": 1, "title": "json-server", "author": "typicode" }
+----
+
+Boom. A mock server.
+
+You can host that server somewhere simple (shove it on Heroku, Amazon EC2,
+etc.), or you can use https://my-json-server.typicode.com/[My JSON Server] - the
+hosted version from the same author.
+
+image::images/mocks-my-json-server.png[]
+
+Now, at this point you are probably thinking: "That's lovely and all, but
+maintaining that one JSON file is going to be a freaking nightmare!"
+
+Yup. Adding a single property to a resource means you are going to have to go
+through each record in `db.json`. Beyond that, formatting the file can turn into
+a pain in the ass, conflicts will be rife as the file grows and different
+developers add different properties, etc. It's good for quick and dirty demos,
+but is probably not something you will want to use in anger.
+
+How about instead of manually maintaining `db.json`, a tool could be pointed at API contracts, and a mock server was automagically created?! Then it figure out its own
```

Comment:
```suggestion
How about instead of manually maintaining `db.json`, a tool could be pointed at API contracts, and a mock server was automagically created?! Then it could figure out its own
```